Field access policy with relation field access cause compile error

```prisma
model User {
    id Int @id @default(autoincrement())
    name String?
    admin Boolean @default(false)

    companiesWorkedFor Company[]

    username String @unique @allow("all", auth() == this) @allow('read', companiesWorkedFor?[owner == auth()]) @allow("all", auth().admin)
}

model Company {
    id Int @id @default(autoincrement())
    name String?
    owner User @relation(fields: [ownerId], references: [id])
    ownerId Int
}
```
After running `zenstack generate`, it will shows the below error:

```bash
node_modules/.zenstack/policy.ts:15:90 - error TS1005: ':' expected.

15                     input, ['id']) || (input?.companiesWorkedFor ? (input?.owner == user)) || (user?.admin ?? null))
```


It misses the `: null` in the collection ex

--- a/packages/schema/src/language-server/validator/expression-validator.ts
+++ b/packages/schema/src/language-server/validator/expression-validator.ts
@@ -115,7 +115,7 @@ export default class ExpressionValidator implements AstValidator<Expression> {
 
             case '==':
             case '!=': {
-                if (expr.left.$resolvedType?.array !== expr.right.$resolvedType?.array) {
+                if (!!expr.left.$resolvedType?.array !== !!expr.right.$resolvedType?.array) {
                     accept('error', 'incompatible operand types', { node: expr });
                     break;
                 }

--- a/packages/schema/src/language-server/validator/expression-validator.ts
+++ b/packages/schema/src/language-server/validator/expression-validator.ts
@@ -64,6 +64,16 @@ export default class ExpressionValidator implements AstValidator<Expression> {
             case '<=':
             case '&&':
             case '||': {
+                if (expr.left.$resolvedType?.array) {
+                    accept('error', 'operand cannot be an array', { node: expr.left });
+                    break;
+                }
+
+                if (expr.right.$resolvedType?.array) {
+                    accept('error', 'operand cannot be an array', { node: expr.right });
+                    break;
+                }
+
                 let supportedShapes: ExpressionType[];
                 if (['>', '>=', '<', '<='].includes(expr.operator)) {
                     supportedShapes = ['Int', 'Float', 'DateTime', 'Any'];
@@ -105,6 +115,11 @@ export default class ExpressionValidator implements AstValidator<Expression> {
 
             case '==':
             case '!=': {
+                if (expr.left.$resolvedType?.array !== expr.right.$resolvedType?.array) {
+                    accept('error', 'incompatible operand types', { node: expr });
+                    break;
+                }
+
                 // disallow comparing model type with scalar type or comparison between
                 // incompatible model types
                 const leftType = expr.left.$resolvedType?.decl;

--- a/packages/schema/tests/schema/validation/attribute-validation.test.ts
+++ b/packages/schema/tests/schema/validation/attribute-validation.test.ts
@@ -559,6 +559,50 @@ describe('Attribute tests', () => {
             model A {
                 id String @id
                 x Int
+                y Int[]
+
+                @@allow(true, x == y)
+            }
+        `)
+        ).toContain('incompatible operand types');
+
+        expect(
+            await loadModelWithError(`
+            ${prelude}
+            model A {
+                id String @id
+                x Int
+                y Int[]
+
+                @@allow(true, x > y)
+            }
+        `)
+        ).toContain('operand cannot be an array');
+
+        expect(
+            await loadModelWithError(`
+            ${prelude}
+            model User {
+                id Int @id
+                foo Foo @relation(fields: [fooId], references: [id])
+                fooId Int
+            }
+
+            model Foo {
+                id Int @id
+                users User[]
+
+                @@allow('all', users == auth())
+            }
+        `)
+        ).toContain('incompatible operand types');
+
+        expect(
+            await loadModelWithError(`
+            ${prelude}
+            model A {
+                id String @id
+                x Int
                 other A? @relation('other', fields: [otherId], references: [id])
                 otherId String? @unique
                 holder A? @relation('other')

